File: autogpt_platform/frontend/src/app/(platform)/copilot/providerFailure.ts
/**
 * The typed reason a provider refused a turn, as it arrives on the stream.
 *
 * Until this existed the client had to sniff substrings out of error text --
 * `"usage limit"`, `"401"` -- to guess what happened, and everything it
 * could not place got the same advice: press Try Again. That advice is
 * wrong for most provider failures. Retrying an expired login, a spent
 * quota or a retired model cannot succeed, and telling someone to retry
 * costs them the time it takes to find out.
 *
 * The server decides `retryable` and `reconnectFixesIt` so a second
 * implementation of that judgement cannot drift from the one that classified
 * the failure.
 */

export type ProviderFailureKind =
  | "auth_expired"
  | "invalid_credential"
  | "usage_limit"
  | "model_unavailable"
  | "policy_denied"
  | "entitlement_required"
  | "transient";

export interface ProviderFailure {
  kind: ProviderFailureKind;
  message: string;
  authProvider: string | null;
  credentialId: string | null;
  resetsAt: number | null;
  retryable: boolean;
  reconnectFixesIt: boolean;
}

export interface ProviderRoute {
  authProvider: string;
  credentialId: string | null;
}

const KINDS = new Set<string>([
  "auth_expired",
  "invalid_credential",
  "usage_limit",
  "model_unavailable",
  "policy_denied",
  "entitlement_required",
  "transient",
]);

/**
 * The envelope, from wherever it came from.
 *
 * The same object rides the live stream and is persisted onto the error
 * marker row, so reopening a chat tomorrow can offer the same recovery the
 * user was offered when it failed. One parser for both, because a second one
 * would be free to disagree about the shape.
 */
export function parseProviderFailure(value: unknown): ProviderFailure | null {
  const data = value as Partial<ProviderFailure> | undefined;
  if (!data || typeof data.kind !== "string" || !KINDS.has(data.kind)) {
    return null;
  }
  return {
    kind: data.kind as ProviderFailureKind,
    message: typeof data.message === "string" ? data.message : "",
    authProvider: data.authProvider ?? null,
    credentialId: data.credentialId ?? null,
    resetsAt: typeof data.resetsAt === "number" ? data.resetsAt : null,
    retryable: data.retryable === true,
    reconnectFixesIt: data.reconnectFixesIt === true,
  };
}

export function parseProviderFailurePart(dataPart: {
  type: string;
  data?: unknown;
}): ProviderFailure | null {
  if (dataPart.type !== "data-provider-failure") return null;
  return parseProviderFailure(dataPart.data);
}

/** The key the backend stores the envelope under on a marker row. */
export const PROVIDER_FAILURE_KEY = "provider_failure";

// Server-written prefixes that mark a row as a failure card rather than a
// reply. Kept in step with COPILOT_ERROR_PREFIX in the backend's constants.
const ERROR_MARKER_PREFIXES = [
  "[__COPILOT_ERROR_f7a1__]",
  "[__COPILOT_RETRYABLE_ERROR_f7a1__]",
];

function isErrorMarker(row: { role?: unknown; content?: unknown }): boolean {
  const content = row.content;
  if (row.role !== "assistant" || typeof content !== "string") return false;
  return ERROR_MARKER_PREFIXES.some((prefix) => content.startsWith(prefix));
}

/**
 * The failure this chat is *currently* sitting on, if any.
 *
 * Scans newest-first and stops at the first thing that answers the question.
 * A marker carrying an envelope is the answer. So is a newer turn: if the user
 * has spoken again, or the assistant has replied for real, since the failure,
 * then the chat moved on and the old failure is history. Without that second
 * stop the scan walks straight past a recovered turn and resurrects a "continue
 * on ..." offer for a failure that no longer applies -- which on a reload is
 * indistinguishable, to the user, from failing again.
 */
export function latestProviderFailure(
  messages: readonly unknown[],
  activeRoute: ProviderRoute | null = null,
): ProviderFailure | null {
  for (let i = messages.length - 1; i >= 0; i--) {
    const row = messages[i];
    if (!row || typeof row !== "object") continue;
    const typed = row as {
      role?: unknown;
      content?: unknown;
      metadata?: unknown;
    };

    if (isErrorMarker(typed)) {
      const meta = typed.metadata;
      if (!meta || typeof meta !== "object") return null;
      const raw = (meta as Record<string, unknown>)[PROVIDER_FAILURE_KEY];
      const failure = raw === undefined ? null : parseProviderFailure(raw);
      if (!failure || !activeRoute || failure.authProvider === null) {
        return failure;
      }
      return failure.authProvider === activeRoute.authProvider &&
        failure.credentialId === activeRoute.credentialId
        ? failure
        : null;
    }

    // A real turn after the failure means the chat recovered.
    if (typed.role === "user") return null;
    if (typed.role === "assistant" && typeof typed.content === "string") {
      return null;
    }
  }
  return null;
}

export function providerFailureFingerprint(failure: ProviderFailure): string {
  return JSON.stringify([
    failure.kind,
    failure.message,
    failure.authProvider,
    failure.credentialId,
    failure.resetsAt,
  ]);
}

interface FailureCopy {
  title: string;
  description: string;
}

const COPY_BY_KIND: Record<ProviderFailureKind, FailureCopy> = {
  auth_expired: {
    title: "Your connection needs signing in again",
    description: "Reconnect the account in Settings, then send this again.",
  },
  invalid_credential: {
    title: "That connection can't be used",
    description: "Reconnect the account in Settings, then send this again.",
  },
  usage_limit: {
    title: "You've hit this connection's limit",
    description: "Switch connection to keep going, or wait for it to reset.",
  },
  model_unavailable: {
    title: "That model isn't available",
    description: "Pick another model tier, or switch connection.",
  },
  policy_denied: {
    title: "The provider declined this request",
    description: "Rewording it may help. Retrying it unchanged won't.",
  },
  entitlement_required: {
    title: "Your plan doesn't include this connection",
    description: "Switch connection, or see plans to unlock it.",
  },
  transient: {
    title: "Connection hiccup",
    description: "Something went wrong on the way to the model. Try again.",
  },
};

/**
 * What to tell the user, preferring the provider's own words for the detail.
 *
 * The reset time is appended only when the provider reported one -- an
 * invented "try again in an hour" is worse than silence, because people plan
 * around it.
 */
export function describeProviderFailure(failure: ProviderFailure): FailureCopy {
  const base = COPY_BY_KIND[failure.kind];
  const reset = formatResetHint(failure.resetsAt);
  const detail = failure.message.trim() || base.description;
  return {
    title: base.title,
    description: reset ? `${base.description} ${reset}` : detail,
  };
}

function formatResetHint(resetsAt: number | null): string | null {
  if (resetsAt === null) return null;
  const seconds = resetsAt - Math.floor(Date.now() / 1000);
  if (seconds <= 0) return null;
  if (seconds < 90) return "Resets in under a minute.";
  const minutes = Math.round(seconds / 60);
  if (minutes < 60) return `Resets in about ${minutes} minutes.`;
  const hours = Math.round(minutes / 60);
  return hours === 1
    ? "Resets in about an hour."
    : `Resets in about ${hours} hours.`;
}

File: autogpt_platform/backend/backend/copilot/provider_failure.py
"""What went wrong with the provider behind a turn, said precisely.

Every provider failure currently reaches the client as one opaque
``baseline_error`` carrying a stringified exception. An expired ChatGPT
login, a monthly usage limit, a retired model and a policy refusal are all
the same event to the UI, so it can only offer the same useless advice --
"Press Try Again" -- to three failures where trying again cannot work.

This module names the difference once, at the boundary where the provider
exception is still intact, and hands the caller something a UI can act on:
whether retrying is pointless, whether reconnecting would help, and when
the limit actually lifts.

``resets_at`` is deliberately ``None`` unless the provider reported a real
timestamp. An invented "try again in an hour" that turns out to be wrong is
worse than saying nothing, because the user plans around it.
"""

import logging
import re
import time
from enum import Enum
from typing import Any

from openai import (
    APIConnectionError,
    APIStatusError,
    APITimeoutError,
    AuthenticationError,
    InternalServerError,
    NotFoundError,
    PermissionDeniedError,
    RateLimitError,
)
from pydantic import BaseModel, Field

from backend.copilot.rate_limit import UserPaywalledError
from backend.integrations.codex.transport import (
    CodexCredentialBusyError,
    CodexCredentialIntegrityError,
    CodexInvocationTimeoutError,
    CodexTransportOverloadedError,
)

logger = logging.getLogger(__name__)


class ProviderFailureKind(str, Enum):
    """Why a provider refused, in terms that imply what to do about it."""

    # The credential was understood and rejected -- reconnecting fixes it.
    AUTH_EXPIRED = "auth_expired"
    # The credential is malformed or unusable; reconnecting fixes it.
    INVALID_CREDENTIAL = "invalid_credential"
    # A quota is spent. Waiting fixes it; retrying now does not.
    USAGE_LIMIT = "usage_limit"
    # The model is gone or not available to this account.
    MODEL_UNAVAILABLE = "model_unavailable"
    # The provider refused on policy grounds. Retrying is pointless.
    POLICY_DENIED = "policy_denied"
    # The user's plan does not include this route.
    #
    # Shares its value with ``ExecutionFailureReason.ENTITLEMENT_REQUIRED``
    # on purpose: the same denial can terminate a graph run or a chat turn,
    # and two names for it would split every downstream count in half.
    ENTITLEMENT_REQUIRED = "entitlement_required"
    # Anything that might genuinely succeed on a second attempt.
    TRANSIENT = "transient"


# Retrying the same turn is only honest for a failure that could resolve
# itself. Everything else needs the user to do something first.
_RETRYABLE = frozenset({ProviderFailureKind.TRANSIENT})

# The kinds a reconnect would actually clear.
_RECONNECTABLE = frozenset(
    {
        ProviderFailureKind.AUTH_EXPIRED,
        ProviderFailureKind.INVALID_CREDENTIAL,
    }
)


class ProviderFailure(BaseModel):
    """One provider refusal, carried whole from the runtime to the client."""

    kind: ProviderFailureKind
    message: str = Field(description="What to tell the user, already humanized")
    auth_provider: str | None = Field(
        default=None, description="Which connection failed: platform or codex"
    )
    credential_id: str | None = None
    resets_at: int | None = Field(
        default=None,
        description=(
            "Unix seconds, only when the provider reported one. Never guessed."
        ),
    )

    @property
    def retryable(self) -> bool:
        return self.kind in _RETRYABLE

    @property
    def reconnect_fixes_it(self) -> bool:
        return self.kind in _RECONNECTABLE

    def as_part(self) -> dict[str, Any]:
        """The payload carried on the stream, including derived advice.

        ``retryable`` is computed here rather than left to the client so a
        second implementation of "is this worth retrying" cannot drift from
        this one.
        """
        return {
            "kind": self.kind.value,
            "message": self.message,
            "authProvider": self.auth_provider,
            "credentialId": self.credential_id,
            "resetsAt": self.resets_at,
            "retryable": self.retryable,
            "reconnectFixesIt": self.reconnect_fixes_it,
        }


# Codex failures arrive carrying an internal code as their message
# ("codex_credential_busy"), which is a log line, not something to show a
# person. These say the same thing in the user's terms, and say what to do.
_CODEX_MESSAGES: dict[type[BaseException], str] = {
    # Deliberately does not say "only one chat at a time". Measured against a
    # live ChatGPT connection, concurrent chats do NOT contend: three
    # simultaneous turns all completed in ~8s, and a short turn fired six
    # seconds into a 42s one answered in 4s, with no busy error in either
    # case. The five-second timeout is on the *credential* lease, which is
    # contended when something needs to write the credential -- a token
    # refresh -- not by ordinary turns sharing it.
    #
    # So this is rarer than "you have two chats open", and wording it that
    # way would send the user to close a chat that is not the problem.
    CodexCredentialBusyError: (
        "This ChatGPT connection is briefly unavailable. " "Try again in a moment."
    ),
    CodexCredentialIntegrityError: (
        "This ChatGPT connection can't be used. Reconnect the account in "
        "Settings, then send this again."
    ),
    CodexTransportOverloadedError: (
        "ChatGPT is busy right now. Try again in a moment."
    ),
    CodexInvocationTimeoutError: ("ChatGPT took too long to respond. Try again."),
}


def _humanize(exc: BaseException) -> str | None:
    """A user-facing sentence for a failure whose own message is a code."""
    return _CODEX_MESSAGES.get(type(exc))


def classify(
    exc: BaseException,
    *,
    auth_provider: str | None = None,
    credential_id: str | None = None,
    message: str | None = None,
) -> ProviderFailure | None:
    """Name the failure, or return ``None`` if it is not the provider's.

    ``None`` means "not recognisably a provider refusal" and leaves the
    caller's existing behaviour untouched. That matters: a bug in our own
    code dressed up as a provider failure would send the user to reconnect
    an account that was never broken.
    """
    kind = _kind_of(exc)
    if kind is None:
        return None
    return ProviderFailure(
        kind=kind,
        message=message or _humanize(exc) or str(exc) or type(exc).__name__,
        auth_provider=auth_provider,
        credential_id=credential_id,
        resets_at=_resets_at(exc),
    )


# HTTP status the compat gateway should answer with for each kind, so the
# CLI upstream of it can tell "stop asking" from "try again". A blanket 502
# reads as a server fault and invites a retry that cannot succeed.
_STATUS_BY_KIND: dict[ProviderFailureKind, int] = {
    ProviderFailureKind.AUTH_EXPIRED: 401,
    ProviderFailureKind.INVALID_CREDENTIAL: 401,
    ProviderFailureKind.ENTITLEMENT_REQUIRED: 402,
    ProviderFailureKind.POLICY_DENIED: 403,
    ProviderFailureKind.MODEL_UNAVAILABLE: 404,
    ProviderFailureKind.USAGE_LIMIT: 429,
    ProviderFailureKind.TRANSIENT: 503,
}


def status_for(failure: "ProviderFailure") -> int:
    """The status a gateway should return for this failure."""
    return _STATUS_BY_KIND.get(failure.kind, 502)


def _kind_of(exc: BaseException) -> ProviderFailureKind | None:
    if isinstance(exc, UserPaywalledError):
        return ProviderFailureKind.ENTITLEMENT_REQUIRED
    # Codex runs behind a CLI, so its failures arrive as transport
    # exceptions rather than HTTP errors. Only the ones whose meaning is
    # unambiguous are named; a bare CodexTransportError could be anything.
    if isinstance(exc, CodexCredentialIntegrityError):
        return ProviderFailureKind.INVALID_CREDENTIAL
    if isinstance(
        exc,
        (
            CodexInvocationTimeoutError,
            CodexTransportOverloadedError,
            CodexCredentialBusyError,
        ),
    ):
        return ProviderFailureKind.TRANSIENT
    if isinstance(exc, AuthenticationError):
        return ProviderFailureKind.AUTH_EXPIRED
    if isinstance(exc, PermissionDeniedError):
        return ProviderFailureKind.POLICY_DENIED
    if isinstance(exc, RateLimitError):
        return ProviderFailureKind.USAGE_LIMIT
    if isinstance(exc, NotFoundError):
        return ProviderFailureKind.MODEL_UNAVAILABLE
    if isinstance(exc, (APIConnectionError, APITimeoutError, InternalServerError)):
        return ProviderFailureKind.TRANSIENT
    if isinstance(exc, APIStatusError):
        # Any other 5xx is worth another attempt; a 4xx we have not named
        # above is a request we built wrong, which is ours, not the
        # provider's, and must not be dressed up as a provider refusal.
        return ProviderFailureKind.TRANSIENT if exc.status_code >= 500 else None
    return None


def _resets_at(exc: BaseException) -> int | None:
    """A reset time only if the provider actually sent one.

    Read from the standard ``x-ratelimit-reset-*`` response headers. Anything
    unparseable is dropped rather than approximated -- a wrong time is worse
    than no time, because the user schedules around it.
    """
    response = getattr(exc, "response", None)
    headers = getattr(response, "headers", None)
    if headers is None:
        return None
    raw = headers.get("x-ratelimit-reset-requests") or headers.get(
        "x-ratelimit-reset-tokens"
    )
    if not raw:
        return None
    try:
        return _to_unix_seconds(str(raw))
    except (ValueError, TypeError):
        logger.debug("Unparseable rate-limit reset header: %r", raw)
        return None


def _to_unix_seconds(raw: str) -> int | None:
    """Absolute epoch seconds, or ``None`` for a shape we do not understand.

    OpenAI sends a duration like ``60s`` / ``1m30s`` / ``6m0s`` / ``2h``;
    some compatible servers send an absolute epoch instead. A duration is
    still the provider reporting a reset -- "60 seconds from now" is a fact,
    not a guess -- so it is anchored to the current clock. What is refused is
    inventing a reset the provider never mentioned at all.
    """
    value = raw.strip().lower()
    if not value:
        return None
    if value.isdigit():
        seconds = int(value)
        # Below this, it is a duration in seconds rather than an epoch:
        # 10**9 seconds is 2001, so no real reset epoch is ever smaller.
        return seconds if seconds > 10**9 else int(time.time()) + seconds
    total = _duration_seconds(value)
    return int(time.time() + total) if total is not None else None


_DURATION_UNITS = {"ms": 0.001, "s": 1.0, "m": 60.0, "h": 3600.0, "d": 86400.0}
_DURATION_PART = re.compile(r"(\d+(?:\.\d+)?)(ms|s|m|h|d)")


def _duration_seconds(value: str) -> float | None:
    """Seconds in a compound duration, or ``None`` if it is not one.

    Whole-string match, so a value carrying anything we did not parse is
    rejected rather than silently half-read.
    """
    parts = _DURATION_PART.findall(value)
    if not parts:
        return None
    if "".join(number + unit for number, unit in parts) != value:
        return None
    return sum(float(number) * _DURATION_UNITS[unit] for number, unit in parts)
